jetty-http2/http2-hpack/src/main/java/org/eclipse/jetty/http2/hpack/HpackEncoder.java

```diff
@@ -142,80 +143,120 @@ public void setLocalMaxDynamicTableSize(int localMaxDynamicTableSize)
+            if (maxDynamicTableSize != _context.getMaxDynamicTableSize())
+                encodeMaxDynamicTableSize(buffer, maxDynamicTableSize);
+
+            // Add Request/response meta fields
+            if (metadata.isRequest())
+            {
+                MetaData.Request request = (MetaData.Request)metadata;
+
+                // TODO optimise these to avoid HttpField creation
+                String scheme = request.getURI().getScheme();
+                encode(buffer, new HttpField(HttpHeader.C_SCHEME, scheme == null ? HttpScheme.HTTP.asString() : scheme));
+                encode(buffer, new HttpField(HttpHeader.C_METHOD, request.getMethod()));
+                encode(buffer, new HttpField(HttpHeader.C_AUTHORITY, request.getURI().getAuthority()));
+                encode(buffer, new HttpField(HttpHeader.C_PATH, request.getURI().getPathQuery()));
+            }
+            else if (metadata.isResponse())
             {
-                if (hopHeaders == null)
-                    hopHeaders = new HashSet<>();
-                hopHeaders.add(StringUtil.asciiToLowerCase(value));
+                MetaData.Response response = (MetaData.Response)metadata;
+                int code = response.getStatus();
+                HttpField status = code < STATUSES.length ? STATUSES[code] : null;
+                if (status == null)
+                    status = new HttpField.IntValueHttpField(HttpHeader.C_STATUS, code);
+                encode(buffer, status);
             }
-            for (HttpField field : fields)
+
+            // Remove fields as specified in RFC 7540, 8.1.2.2.
+            if (fields != null)
             {
-                HttpHeader header = field.getHeader();
-                if (header != null && IGNORED_HEADERS.contains(header))
-                    continue;
-                if (header == HttpHeader.TE)
+                // For example: Connection: Close, TE, Upgrade, Custom.
+                Set<String> hopHeaders = null;
+                for (String value : fields.getCSV(HttpHeader.CONNECTION, false))
                 {
-                    if (field.contains("trailers"))
-                        encode(buffer, TE_TRAILERS);
-                    continue;
+                    if (hopHeaders == null)
+                        hopHeaders = new HashSet<>();
+                    hopHeaders.add(StringUtil.asciiToLowerCase(value));
+                }
+                for (HttpField field : fields)
+                {
+                    HttpHeader header = field.getHeader();
+                    if (header != null && IGNORED_HEADERS.contains(header))
+                        continue;
+                    if (header == HttpHeader.TE)
+                    {
+                        if (field.contains("trailers"))
+                            encode(buffer, TE_TRAILERS);
+                        continue;
+                    }
+                    String name = StringUtil.asciiToLowerCase(field.getName());
```

Comment:
Can we come up with a way to query HttpHeader instances for their lower case names and only do a real asciiToLowerCase for unknown headers.... and then we should only do any of this IFF we have hopHeaders